lib/sles4sap/console_redirection.pm

```diff
@@ -0,0 +1,170 @@
+    croak "OpenQA variable WORKER_VM_ID undefined. Run 'redirection_init()' first" unless get_var('WORKER_VM_ID');
+    croak "IP address '$args{'target_ip'}' is not valid." unless grep(/^$RE{net}{IPv4}$/, $args{'target_ip'});
+    croak 'Global variable "$searialdev" undefined' unless $serialdev;
+    croak "Console is already redirected to:" . script_output('hostname') if check_serial_redirection();
+
+    enter_cmd "ssh $ssh_opt $args{'ssh_user'}\@$args{'target_ip'} 2>&1 | tee -a /dev/$serialdev";
+    handle_login_prompt($args{'ssh_user'});
+    check_serial_redirection();
+    record_info('Redirect ON', "Serial redirection established to: $args{'target_ip'}");
+}
+
+=head2 disconnect_target_from_serial
+
+    disconnect_target_from_serial([worker_machine_id=$worker_machine_id]);
+
+B<worker_machine_id>: Target host IP - default value is defined by OpenQA parameter WORKER_VM_ID from redirect_init()
+
+Disconnects target from serial console by typing 'exit' command until host machine ID matches ID of the worker VM.
+
+=cut
+
+sub disconnect_target_from_serial {
+    my (%args) = @_;
+    $args{worker_machine_id} //= get_required_var('WORKER_VM_ID');
+    set_serial_term_prompt();
+    my $serial_redirection_status = check_serial_redirection(worker_machine_id => $args{worker_machine_id});
+    while ($serial_redirection_status ne 0) {
+        enter_cmd('exit');    # Enter command and wait for screen start changing
+        $testapi::distri->{serial_term_prompt} = '';    # reset console prompt
+        wait_serial(qr/Connection.*closed./, timeout => 10);    # Wait for connection to close
+        wait_serial(qr/# |> /, timeout => 10);    # Wait for console prompt to appear
+        set_serial_term_prompt();    # after logout user might change and prompt with it.
+        $serial_redirection_status = check_serial_redirection($args{worker_machine_id});
+    }
+    record_info('Redirect OFF', "Serial redirection closed. Console set to: " . script_output('hostname'));
+}
+
+=head2 check_serial_redirection
+
+    check_serial_redirection([worker_machine_id=$worker_machine_id]);
+
+B<worker_machine_id>: Target host IP - default value is defined by OpenQA parameter WORKER_VM_ID from redirect_init()
+
+Compares current machine-id to the worker VM ID either defined by WORKER_VM_ID variable or positional argument.
+Machine ID is used instead of IP addr since cloud VM IP might not be visible from the inside (for example via 'ip a')
+
+=cut
+
+sub check_serial_redirection {
+    my (%args) = @_;
+    $args{worker_machine_id} //= get_required_var('WORKER_VM_ID');
+    my $current_id = script_output 'cat /etc/machine-id';
+
+    my $redirection_status = $current_id eq $args{worker_machine_id} ? 0 : 1;
+    record_info('Redir check', 'Console redirection is not active') unless $redirection_status;
+    record_info('Redir check', "Console is redirected to: " . script_output('hostname')) if $redirection_status;
```

Comment:
Suggestion:
```
my $logmsg = 'Console redirection is not active';
$logmsg = 'Console is redirected to: ' . script_output('hostname') if $redirection_status;
record_info('Redir check, $msg);
```
You can also use the ternary operator in `record_info`, but I think it's not as legible:
```
record_info('Redir check', ($redirection_status ?
    'Console is redirected to: ' . script_output('hostname') : 'Console redirection is not active'));
```

Reply:
```
    my $logmsg = $redirection_status ? 'Console is redirected to: ' . script_output('hostname') : 'Console redirection is not active';
    record_info('Redir check', $logmsg);
```
I smashed them together !